versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/indexes/StoreIndexImpl.java

```diff
@@ -517,31 +564,41 @@ private StoreIndexImpl(ByteBuffer serialized, ElementSerializer<V> ser) {
         version >= 2 ? new ArrayList<>(readVarInt(serialized)) : new ArrayList<>();
 
     boolean first = true;
-
-    // This buffer holds the previous key, reused.
-    ByteBuffer previousKey = newKeyBuffer();
+    int previousKeyLen = 0;
+    LazyStoreIndexElement predecessor = null;
+    LazyStoreIndexElement previous = null;
 
     while (serialized.remaining() > 0) {
       int strip = first ? 0 : readVarInt(serialized);
       first = false;
 
-      // strip
-      previousKey.position(previousKey.position() - strip);
-      previousKey.limit(MAX_KEY_BYTES);
-      // add
-      int limitSave = serialized.limit();
-      previousKey.put(serialized.limit(findPositionAfterKey(serialized)));
-      serialized.limit(limitSave);
-      // read key
-      previousKey.flip();
-      StoreKey key = StoreKey.deserializeKey(previousKey);
-
+      int prefixLen = previousKeyLen - strip;
+      int keyOffset = serialized.position();
+      skipKey(serialized); // skip key
       int valueOffset = serialized.position();
-      ser.skip(serialized);
+      ser.skip(serialized); // skip content/value
       int endOffset = serialized.position();
 
-      LazyStoreIndexElement element = new LazyStoreIndexElement(key, valueOffset, endOffset);
+      int keyPartLen = valueOffset - keyOffset;
+      int totalKeyLen = prefixLen + keyPartLen;
+
+      predecessor = cutPredecessor(predecessor, prefixLen, previous);
```

Comment:
Is it worth performing this optimization at index creation time? Couldn't we do the same thing at key materialization time (storing intermediate traversal results in `LazyStoreIndexElement`), which may or may not happen for every element?

Reply:
At materialization time, it "walks backwards" already using the available `predecessor` - at deserialization time, it "walks forwards".

At deserialization time it is much cheaper, because it does not have the super long tail of predecessors that it would have without `cutPredecessor`.

Reply:
I mean that we could execute the logic of `cutPredecessor` at materialization time. For a given key we have to trace the same sequence of previous elements to find the best predecessor regardless of when we do this traversal. The first time we find an area, which we can skip, we can record that in the element to avoid traversing the skipped segment again on new meterializations.

However, this is probably a very minor point. Totally optional.

Reply:
Well, it would have to do the same "cut" for all key, whereas deserialization does it implicitly for all "following keys"